beacon_node/beacon_chain/src/attestation_rewards.rs

```diff
@@ -0,0 +1,187 @@
+
+            let unslashed_participating_increments =
+                unslashed_participating_balance.safe_div(spec.effective_balance_increment)?;
+
+            let total_active_balance = participation_cache.current_epoch_total_active_balance();
+
+            let active_increments =
+                total_active_balance.safe_div(spec.effective_balance_increment)?;
+
+            let base_reward_per_increment =
+                BaseRewardPerIncrement::new(total_active_balance, spec)?;
+
+            for effective_balance_eth in 0..=32 {
+                let base_reward =
+                    effective_balance_eth.safe_mul(base_reward_per_increment.as_u64())?;
+
+                let penalty = -(base_reward.safe_mul(weight)?.safe_div(WEIGHT_DENOMINATOR)? as i64);
+
+                let reward_numerator = base_reward
+                    .safe_mul(weight)?
+                    .safe_mul(unslashed_participating_increments)?;
+
+                let ideal_reward = reward_numerator
+                    .safe_div(active_increments)?
+                    .safe_div(WEIGHT_DENOMINATOR)?;
+                if !state.is_in_inactivity_leak(previous_epoch, spec) {
+                    ideal_rewards_hashmap
+                        .insert((flag_index, effective_balance_eth), (ideal_reward, penalty));
+                } else {
+                    ideal_rewards_hashmap.insert((flag_index, effective_balance_eth), (0, penalty));
+                }
+            }
+        }
+
+        // Calculate total rewards
+        let mut total_rewards: Vec<TotalAttestationRewards> = Vec::new();
+
+        let penalty = 0;
+
+        if validators.is_empty() {
+            validators = participation_cache.eligible_validator_indices().to_vec();
+        }
+
+        for validator_index in &validators {
+            let eligible = state.is_eligible_validator(previous_epoch, *validator_index)?;
+
+            let effective_balance = state.get_effective_balance(*validator_index)?;
+
+            let effective_balance_eth =
+                effective_balance.safe_div(spec.effective_balance_increment)?;
+
+            let mut head_reward = 0u64;
+            let mut target_reward = 0i64;
+            let mut source_reward = 0i64;
+
+            for flag_index in 0..PARTICIPATION_FLAG_WEIGHTS.len() {
+                if eligible {
+                    let voted_correctly = participation_cache
+                        .get_unslashed_participating_indices(flag_index, previous_epoch)
+                        .map_err(|_| BeaconChainError::AttestationRewardsError)?
+                        .contains(*validator_index)
+                        .map_err(|_| BeaconChainError::AttestationRewardsError)?;
+                    if voted_correctly {
+                        let (ideal_reward, _penalty) = ideal_rewards_hashmap
+                            .get(&(flag_index, effective_balance_eth))
+                            .ok_or(BeaconChainError::AttestationRewardsError)?;
```

Comment:
If you raise this out of the `if voted_correctly { .. }` you can use the `penalty` in the `else if { }` branches